readme.md
# Averages extension for weewx #

## Description ##

The Averages extension for `weewx` calculates temperature maxima and minima as well as temperature and rainfall  averages for each month and generates a JSON format file suitable for plotting this data using Highcharts. A sample HTML page and the JavaScript necessary to render a Highcharts plot of the resulting data are included with the Averages extension.

An example plot produced using the Averages extension is included below:

![Example monthly temperature and rainfall averages plot](https://github.com/gjr80/weewx-averages/blob/master/chart.png "Example monthly temperature and rainfall averages plot")

The Averages extension consists of a `weewx` Search List Extension (SLE) that calculates the monthly aggregates, a skin that generates JSON format data file containing the monthly data and a sample HTML page and supporting JavaScript that displays the monthly temperature and rainfall data on a Highcharts generated graph.


## Pre-Requisites ##

The Averages extension requires `weewx v3.0.0` or greater. Some optional features of the Averages extension require `weewx 3.6.0` or greater. Refer to the [User's Guide](https://github.com/gjr80/weewx-averages/wiki/User's-Guide "Averages extension User's Guide").  

## Installation ##

Most user's should install the Averages extension using the [Quick-Start Guide](https://github.com/gjr80/weewx-averages/wiki/Quick-Start-Guide "Averages extension Quick-Start Guide"). Further details along with manual installation, uninstallation and customization instructions can be found in the [User's Guide](https://github.com/gjr80/weewx-averages/wiki/User's-Guide "Averages extension User's Guide").

## Support ###

General support issues may be raised in the Google Groups [weewx-user forum](https://groups.google.com/group/weewx-user "Google Groups weewx-user forum"). Specific bugs in the Averages extension code should be the subject of a new issue raised via the [Issues Page](https://github.com/gjr80/weewx-averages/issues "Averages extension Issues").
 
## Licensing ##

The Averages extension for weewx is licensed under the [GNU Public License v3](https://github.com/gjr80/weewx-averages/blob/master/LICENSE "Averages extension License").

changelog
v1.0.0
*   compatible with WeeWX 4.0.0 under python 2 or python 3
v0.5.0
*   initial implementation (started at v0.5.0 to align with SLE version)
*   now packaged as a weewx extension
*   reworked the SLE class:
    -   took get_first_day() method out of the SLE class
    -   consolidated SLE getMonthAveragesHighsLows() method into 
        get_extension_list() method
    -   reworked algorithm to calculate means, now use BoM definitions. Refer 
        class comments for details.
    -   no longer use partial months at start and end of archive for anything, 
        only use complete archive months. Note, any partial months after the 
        first complete month and before the last complete month are still 
        'included'.
*   added config{} to averages.js allow users to change common look and feel
    settings without delving into the code.


skins/HighchartsAverages/scripts/theme.js
/**
 * Theme distributed with Averages weewx extension
 * Based on Gray theme for Highcharts JS by Torstein Honsi
 **/

Highcharts.theme = {
    chart: {
        backgroundColor: 'rgb(208, 208, 208)',
        borderWidth: 1,
        borderColor: '#000000',
        borderRadius: 8,
        plotShadow: false,
        plotBorderWidth: 0,
    },

    colors: ["#B44242", "#4242B4", "#42B442", "#DF5353", "#aaeeee", "#ff0066", "#eeaaee",
        "#55BF3B", "#DF5353", "#7798BF", "#aaeeee"],

    labels: {
        style: {
            color: '#CCC'
        }
    },

    legend: {
        borderWidth: 0,
        itemStyle: {
            color: '#555'
        },
        itemHoverStyle: {
            color: '#FFF'
        },
        itemHiddenStyle: {
            color: '#999'
        },
        margin: 5,
        padding: 4,
        symbolPadding: 2
    },

    plotOptions: {
        column: {
            shadow: false
        },
    },

    subtitle: {
        style: {
            color: '#555',
        }
    },

    title: {
        margin: 5,
        style: {
            color: '#555',
            font: '16px Lucida Grande, Lucida Sans Unicode, Verdana, Arial, Helvetica, sans-serif'
        }
    },

    xAxis: {
        labels: {
            style: {
                color: '#555',
                fontWeight: 'bold',
                whiteSpace: 'nowrap'
            }
        },
        minorGridLineWidth: 0,
        minorTickInterval: 'auto',
        minorTickLength: 10,
        tickColor: '#555',
        title: {
            style: {
                color: '#555',
            }
        }
    },
    yAxis: {
        allowDecimals: false,
        gridLineColor: '#AAA',
        gridLineWidth: 1,
        labels: {
            style: {
                color: '#555',
            }
        },
        lineWidth: 1,
        minorTickColor: '#555',
        minorTickInterval: 'auto',
        minorTickLength: 2,
        minorTickWidth: 1,
        minTickInterval: null,
        tickWidth: 1,
        title: {
            style: {
                color: '#555',
                font: 'bold 12px Lucida Grande, Lucida Sans Unicode, Verdana, Arial, Helvetica, sans-serif'
            }
        }
    },

    toolbar: {
        itemStyle: {
            color: '#CCC'
        }
    },

    tooltip: {
        backgroundColor: 'rgba(255, 255, 204, .7)',
        borderWidth: 0,
        style: {
            color: '#555'
        }
    },

    navigation: {
        buttonOptions: {
            symbolStroke: '#DDDDDD',
            hoverSymbolStroke: '#FFFFFF',
            theme: {
                fill: {
                    linearGradient: { x1: 0, y1: 0, x2: 0, y2: 1 },
                    stops: [
                        [0.4, '#606060'],
                        [0.6, '#333333']
                    ]
                },
                stroke: '#000000'
            }
        }
    },
};

// Apply the theme
var highchartsOptions = Highcharts.setOptions(Highcharts.theme);


skins/HighchartsAverages/scripts/averages.js
/**
*
* Copyright (c) 2015-2019 Gary Roderick <gjroderick(at)gmail.com>
*
* Released under GNU General Public License, Version 3, 29 June 2007.
* Refer to the enclosed License file for your full rights.
*
*
* Javascript code to initialise and render WeeWX monthly averages plot using
* Highcharts.
*
*
* Version: 1.0.0a1                                 Date: 30 December 2019
*
* Revision History
*   30 December 2019    v1.0.0
*       - version number change only
*   30 September 2016   v0.5.0
*       - now packaged as a weewx extension
*       - added config to allow users to change common look and feel
*         settings without delving into the code
*  10 May 2016          v0.4.0
*       - no change, version number upgrade only
*  April 2016           v0.3.0
*       - no change, version number upgrade only
*  5 March 2016         v0.2.2
*       - no change, version number upgrade only
*  21 July 2015         v0.2.1
*       - remove old redundant code
*  19 March 2015        v0.2.0
*       - no change, version number upgrade only
*  22 February 2015     v0.1.0
*       - initial implementation
*
**/

var config = {
    /**
    *
    * Below are some options the user may wish to tweak to set the look and
    * feel of the plot. All control various options detailed in the Highcharts
    * API Reference.
    *
    **/

    json_source: 'json/averages.json',  // path to the JSON file holding the source data
    render_to: 'monthaveragesplot',      // id of the HTML element where the chart will be rendered
    title: 'Monthly Temperature and Rainfall Averages',  // plot title. String
    show_legend: true,                      // display plot legend. true|false
    av_temp_range_label: 'Mean Temp Range', // legend label for mean temperature range plot. String
    av_temp_range_color: '#CC3399',         // color for mean temperature range plot. String, color name or RGB
    av_temp_range_opacity: 0.4,             // opacity for mean temperature range plot. Number
    av_temp_label: 'Mean Temp',             // legend label for mean temperature plot. String
    av_temp_color: '#BA55D3',               // color for mean temperature plot. String, color name or RGB
    max_temp_label: 'Max Temp',             // legend label for max temperature plot. String
    max_temp_color: '#FF0000',              // color for max temperature plot. String, color name or RGB
    min_temp_label: 'Min Temp',             // legend label for min temperature plot. String
    min_temp_color: '#0000FF',              // color for min temperature plot. String, color name or RGB
    avg_rainfall_label: 'Avg Rainfall',     // legend label for average rainfall plot. String
    avg_rainfall_color: '#72B2C4',          // color for avg rainfall plot. String, color name or RGB
    background_color_stop1: '#FCFFC5',      // 1st color to be used in background gradient. String, color name or RGB
    background_color_stop2: '#E0E0FF',      // 2nd color to be used in background gradient. String, color name or RGB
    marker_symbol: 'circle',                // marker symbol to be used for each point of each plot (except rainfall). String
    marker_enabled: false,                  // enable marker symbol for each plot(except rainfall). true|false
    updated_align: 'right',                 // alignment of 'Updated: ' label. 'right'|'left'
    updated_font_size: '10px',              // font size of 'Updated: ' label. String
    updated_x_offset: -25,                  // x offset of 'Updated: ' label. Number
    // month names to display on x-axis
    months: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
    x_axis_line_color: '#555',              // x-axis line color. String, color name or RGB
    x_axis_line_width: 1,                   // x-axis line width in pixels. Nubmer
    x_axis_title_color: '#555',             // x-axis title color. String, color name or RGB
    // x-axis font
    x_axis_title_font: 'bold 12px Lucida Grande, Lucida Sans Unicode, Verdana, Arial, Helvetica, sans-serif',
    y_axis_line_color: '#555',              // y-axes line color. String, color name or RGB
    y_axis_line_width: 1,                   // y-axes line width in pixels. Nubmer
    x_axis_title_color: '#555',             // y-axes title color. String, color name or RGB
    // y-axes font
    y_axis_title_font: 'bold 12px Lucida Grande, Lucida Sans Unicode, Verdana, Arial, Helvetica, sans-serif',
    enable_tooltip: true,                   // display tooltip. true|false
    tooltip_font_size: '10px',              // tooltip font size. String
};

/**
*
* Edit anything below here at your own risk; get it wrong and your plot may not
* display and you will need to troubleshoot the JS yourself. All parameters are
* as described in the Highcharts API reference.
*
**/

$(document).ready(function() {

    var optionsAverages = {
        chart: {
            plotBackgroundColor: {
                linearGradient: { x1: 0, y1: 0, x2: 1, y2: 1 },
                stops: [
                    [0, config.background_color_stop1],
                    [1, config.background_color_stop1]
                ]
            },
            renderTo: config.render_to
        },
        legend: {
            enabled: config.show_legend,
            symbolHeight: 12,
            symbolRadius: 0,
            symbolWidth: 12,
        },
        plotOptions: {
            areasplinerange: {
                lineWidth: 0,
                marker: {
                    enabled: config.marker_enabled,
                    radius: 1,
                    symbol: config.marker_symbol
                },
                tooltip: {
                    valueSuffix: ''
                },
            },
            column: {
                borderWidth: 0,
                tooltip: {
                    valueSuffix: ''
                },
            },
            spline: {
                lineWidth: 1,
                marker: {
                    enabled: config.marker_enabled,
                    radius: 1,
                    symbol: config.marker_symbol
                },
                tooltip: {

                    valueSuffix: ''
                },
            },
        },
        series: [{
            name: config.av_temp_range_label,
            type: 'areasplinerange',
            color: config.av_temp_range_color,
            fillOpacity: config.av_temp_range_opacity,
            zIndex: 4,
        }, {
            name: config.av_temp_label,
            type: 'spline',
            color: config.av_temp_color,
            zIndex: 3,
        }, {
            name: config.max_temp_label,
            type: 'spline',
            color: config.max_temp_color,
            zIndex: 2,
        }, {
            name: config.min_temp_label,
            type: 'spline',
            color: config.min_temp_color,
            zIndex: 1,
        }, {
            name: config.avg_rainfall_label,
            type: 'column',
            color: config.avg_rainfall_color,
            zIndex: 0,
            yAxis: 1
        }],
        subtitle: {
            align: config.updated_align,
            style: {
                fontSize: config.updated_font_size
            },
            text: '',
            x: config.updated_x_offset
        },
        title: {
            text: config.title
        },
        tooltip: {
            crosshairs: [true, false],
            enabled: config.enable_tooltip,
            shared: true,
            style: {
                fontSize: config.tooltip_font_size,
            },
            // need to initialise valueSuffix now so we can set it later
            valueSuffix: ''
        },
        xAxis: {
            categories: config.months,
            lineColor: config.x_axis_line_color,
            lineWidth: config.x_axis_line_width,
            title: {
                style: {
                    color: config.x_axis_title_color,
                    font: config.x_axis_title_font
                }
            },
        },
        yAxis: [{
            endOnTick: true,
            lineColor: config.y_axis_line_color,
            lineWidth: config.y_axis_line_width,
            minorGridLineWidth: 0,
            showLastLabel: true,
            startOnTick: true,
            title: {
                style: {
                    color: config.x_axis_title_color,
                    font: config.x_axis_title_font
                },
                // need to initialise text now so we can set it later
                text: ''
            },
        },{
            lineColor: config.y_axis_line_color,
            lineWidth: config.y_axis_line_width,
            title: {
                // need to initialise text now so we can set it later
                text: ''
            },
            opposite: true
        }]
    };

    $.getJSON(config.json_source, function(seriesData) {
        optionsAverages.series[0].data = seriesData[0].temperatureplot.series.outTempMeanMinMax.data;
        optionsAverages.series[1].data = seriesData[0].temperatureplot.series.outTempMean.data;
        optionsAverages.series[2].data = seriesData[0].temperatureplot.series.outTempMax.data;
        optionsAverages.series[3].data = seriesData[0].temperatureplot.series.outTempMin.data;
        optionsAverages.series[4].data = seriesData[0].rainplot.series.rainAvg.data;
        optionsAverages.yAxis[0].title.text = 'Temperature ' + seriesData[0].temperatureplot.yAxisLabel.text;
        optionsAverages.yAxis[1].title.text = 'Rainfall ' + seriesData[0].rainplot.yAxisLabel.text;
        optionsAverages.plotOptions.areasplinerange.tooltip.valueSuffix = seriesData[0].temperatureplot.yAxisUnits.text;
        optionsAverages.plotOptions.spline.tooltip.valueSuffix = seriesData[0].temperatureplot.yAxisUnits.text;
        optionsAverages.plotOptions.column.tooltip.valueSuffix = seriesData[0].rainplot.yAxisUnits.text;
        optionsAverages.subtitle.text = 'Updated: ' + seriesData[0].generated;
        var chart = new Highcharts.Chart(optionsAverages);
    });
});

install.py
"""install.py
Copyright (c) 2015-2016 Gary Roderick <gjroderick(at)gmail.com>

Released under GNU General Public License, Version 3, 29 June 2007.
Refer to the enclosed License file for your full rights.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
details.

Installer for the WeeWX Averages extension.

Version: 1.0.0a1                                   Date: 30 December 2019

Revision History
    30 December 2019
        - removed code that used either report_timing or stale_age depending on
          WeeWX version
    30 September 2016   v0.5.0
        - initial implementation (bumped to v0.5.0 to align with supporting
          SLE version)
"""
import weewx

from distutils.version import StrictVersion
from setup import ExtensionInstaller

# TODO. Fix before release
REQUIRED_VERSION = "4.0.0b5"
AFW_VERSION = "1.0.0"


def loader():
    return AveragesInstaller()


class AveragesInstaller(ExtensionInstaller):
    def __init__(self):
        if StrictVersion(weewx.__version__) < StrictVersion(REQUIRED_VERSION):
            msg = "%s requires WeeWX %s or greater, found %s" % ('Averages ' + AFW_VERSION,
                                                                 REQUIRED_VERSION,
                                                                 weewx.__version__)
            raise weewx.UnsupportedFeature(msg)
        super(AveragesInstaller, self).__init__(
            version="1.0.0a1",
            name='Averages',
            description='Highcharts plots of WeeWX monthly averages.',
            author="Gary Roderick",
            author_email="gjroderick(at)gmail.com",
            config={
                'StdReport': {
                    'HighchartsAverages': {
                        'skin': 'HighchartsAverages',
                        'report_timing': '@monthly',
                        'Units': {
                            'Groups': {
                                'group_rain':        'mm',
                                'group_temperature': 'degree_C',
                            },
                            'StringFormats': {
                                'degree_C': '%.1f',
                                'degree_F': '%.1f',
                                'cm':       '%.2f',
                                'inch':     '%.2f',
                                'mm':       '%.1f'
                            },
                            'TimeFormats': {
                                'current': '%-d %B %Y %H:%M'
                            }
                        },
                        'CopyGenerator': {
                            'copy_once': ['scripts/averages.js',
                                          'scripts/theme.js', 
                                          'averages.html']
                        },
                        'Generators': {
                            'generator_list': ['weewx.cheetahgenerator.CheetahGenerator', 
                                               'weewx.reportengine.CopyGenerator']
                        }
                    }
                }
            },
            files=[('bin/user',                         ['bin/user/averagesSearchX.py']),
                   ('skins/HighchartsAverages',         ['skins/HighchartsAverages/averages.html',
                                                         'skins/HighchartsAverages/skin.conf']),
                   ('skins/HighchartsAverages/json',    ['skins/HighchartsAverages/json/averages.json.tmpl']),
                   ('skins/HighchartsAverages/scripts', ['skins/HighchartsAverages/scripts/averages.js',
                                                         'skins/HighchartsAverages/scripts/theme.js'])
                   ]
        )
